src/ripple/app/main/Application.cpp

```diff
@@ -1110,7 +1064,30 @@ class ApplicationImp : public Application, public RootStoppable, public BasicApp
                 return validators().trustedPublisher(pubKey);
             });
 
-        stopped();
+        // The order of these stop calls is delicate.
+        // Re-ordering them risks undefined behavior.
```

Comment:
I share @cjcoob23's concern:

>I feel that this situation is exactly the problem that Stoppable was meant to solve (managing the order in which things are stopped); with this PR, now we have to manually maintain this order.

This commit reinforces that. I'm not saying that this cleanup and the explicit documentation of the ordering (instead of the implicit and opaque emergent ordering imposed by `Stoppable`) isn't an improvement.

But this makes it hard to figure out where something new should go, or to even be sure that we properly stop everything.

Nothing actionable; food for thought.

Reply:
Here's my idea:

The new way is RAII. (This is actually the 30 years old way, but it sadly hasn't seemed to have caught on yet.) Start your object in its constructor, and stop it in its destructor. Or if you insist on having `start` and `stop` methods, either (1) have `start` return an object that calls `stop` in its destructor or (2) use a separate object, a la `std::lock_guard`, that calls `start` in its constructor and `stop` in its destructor, and never call `start` or `stop` yourself. If you need to dynamically end a lifetime early, own the RAII object via `std::unique_ptr` or `std::optional` (depending on how you want to manage storage) and clear it when you want to stop.

Most objects depend on objects that do not depend on it (no circular dependencies). They should be started after their dependencies and stopped before them. This matches perfectly with the behavior of RAII.

I'd like to make these changes in our codebase, but I expect they will be very large changes. I'd like to start with this changeset as the new baseline.

If it comforts you to know, the order isn't as delicate as the comment makes it seem; I just don't want people touching it. Add your new stoppable objects around it. Start yours after everything else in `ApplicationImp::start`, and stop it before everything else in `ApplicationImp::stop`. If we assume that this order properly stops everything, then changes like that will preserve that invariant.

To prove that this order is correct, I need to demarcate the dependency edges between the components. Part of why this is so difficult is that everyone likes to pull in the kitchen sink (`Application`) even if they only need one or two dependencies. Another part is that we like to build huge monolithic classes with 15 responsibilities and 10 states, with some dependencies active in only a subset of those responsibilities and states. I'd like to get to the point where we know the exact dependencies of each component. That will make it easy to know the correct start and stop sequences (among other benefits).

Reply:
We have a new component `LedgerReplayer`. So I have been thinking the rules of inserting a new component `x` to the list. I'd like to confirm that `x->stop();` can be a blocking call and can call `join()` if `xyz` has threads. And when `x->stop();` returns, if `x` can guarantee that (1) threads are joined if any, (2) will treat calls from other components as no-ops (including callbacks such as a timeout), and to be safe (3) will not call into other components, then `x`'s position in the list does not matter. Right?

Reply:
After `stop` has returned, the object must be in a "stopped" state. A stopped object can be safely destroyed without waiting for anything else. (Short answer: yes it can block, and yes it can join threads.)

A stopped object is not required to support any operations other than destruction. It may choose to allow other operations on a case-by-case basis, and it may choose to implement them as no-ops, but neither is required. Other objects that want to interact with a stopped object must ensure that the stopped object allows those operations while stopped. Let us call these operations "safe operations", and operations that may only be called while the object is in the started state "unsafe operations".

If stopped object B depends on stopped object A, then B must not call any unsafe operations on A after B has stopped.

To answer your last question, we need to add one more requirement: because `x` may call unsafe operations _before_ `x->stop()` has returned, `x->stop()` must be called _before_ any of its dependencies with unsafe operations are stopped. `x` can always be safely placed first in the stop sequence, but cannot be placed after any of its dependencies on whom it calls unsafe operations.

Reply:
Thanks for the explanation. I agree. My concern is that sometimes circular dependencies and safe operations vs unsafe operations are that obvious. Taking a step back, `LedgerMaster` uses `LedgerReplayer` (for the publish use case), and `LedgerReplayer` uses `LedgerMaster` (to check if some data is available locally). `LedgerReplayer` uses `Peer` to send requests, and `Peer` uses `LedgerReplayer` to process the responses. Are those circular dependencies? It was not very clear to me. So we made all the calls and callbacks to `LedgerReplayer` "stop-safe".
I think the developer of a new component should make an effort to make all the interaction with other components (the service it provides and the callbacks from the services it has requested) "stop-safe". The few exceptions left will sort the stop list.

Reply:
`LedgerMaster` should never have been a `Stoppable`, and it is no longer a stoppable object after this changeset. `PeerImp` has never been a stoppable object. Thus, in the two examples you gave, there are no circular stoppable dependencies.

If there ever does arise a circular dependency, only one object in the cycle needs to have safe operations to permit a definitively safe stop sequence. I think we'll find in practice that most stoppable objects don't need to worry about it.